src/core/ts/background-scripts/background.js/telemetry/Telemetry.ts

```diff
@@ -5,75 +5,190 @@
     const appId = config.telemetryAppId;
     Glean.initialize(appId, true, {
       debug: { logPings: config.telemetryDebugMode },
     });
-    this._metricsToSubmit = 0;
     console.info(
       `Telemetry: initialization completed with application ID ${appId}.`,
     );
   }
 
+  public onSelectTranslateFrom(tabId: number) {
+    changeLang.record();
+    this.submit();
+  }
+
+  public onSelectTranslateTo(tabId: number) {
+    changeLang.record();
+    this.submit();
+  }
+
+  public onInfoBarClosed(tabId: number) {
+    closed.record();
+    this.submit();
+  }
+
+  public onNeverTranslateThisSite(tabId: number) {
+    neverTranslateSite.record();
+    this.submit();
+  }
+
+  public onTranslateButtonPressed(tabId: number, from: string, to: string) {
+    fromLang.set(from);
+    toLang.set(to);
+    translate.record();
+    this.submit();
+  }
+
+  public onShowOriginalButtonPressed(tabId: number) {
+    // TODO?
+  }
+
+  public onShowTranslatedButtonPressed(tabId: number) {
+    // TODO?
+  }
+
   /**
-   * Provides access to Telemetry singleton.
+   * A translation attempt starts when a translation is requested in a
+   * specific tab and ends when all translations in that tab has completed
    */
-  public static get global(): Telemetry {
-    if (Telemetry._instance === null) {
-      Telemetry._instance = new Telemetry();
+  public async onTranslationAttemptConcluded(
+    from: string,
+    to: string,
+    $modelLoadTime: number,
+    $translationTime: number,
+    $wordsPerSecond: number,
+    beforeSubmit: () => Promise<void> = undefined,
+  ) {
+    fromLang.set(from);
+    toLang.set(to);
+    modelLoadTime.set(String($modelLoadTime));
+    translationTime.set(String($translationTime));
+    wordsPerSecond.set(String($wordsPerSecond));
+    if (beforeSubmit) {
+      await beforeSubmit();
     }
-    return Telemetry._instance;
+    this.submit();
   }
 
   /**
-   * Enables uploading Glean metrics to telemetry server.
-   * @param val bool
+   * Submits all collected metrics in a custom ping.
    */
-  public setUploadEnabled(val: boolean) {
-    Glean.setUploadEnabled(val);
+  public submit = () => {
+    try {
+      // TODO: Always include the fx telemetry id string metric in pings
+      custom.submit();
+      console.info("Telemetry: the ping is submitted.");
+    } catch (err) {
+      // telemetry error shouldn't crash the app
+      console.error(`Telemetry: Error. The ping was not submitted.`, err);
+    }
+  };
+}
+
+type GleanTransaction = () => Promise<void>;
+
+/* eslint-disable no-unused-vars, no-shadow */
+// TODO: update typescript-eslint when support for this kind of declaration is supported
+interface PendingGleanTransaction<T> {
+  resolve: (T) => void;
+}
+/* eslint-enable no-unused-vars, no-shadow */
+
+interface GleanRecording {
+  gleanTransaction: GleanTransaction;
+  id?: string;
+}
+
+/**
+ * Class responsible for recording Glean telemetry synchronously
+ */
+export class GleanRecorder {
```

Comment:
I noticed this structure is very similar to the [`TranslationRequestManager`](https://github.com/mozilla-extensions/bergamot-browser-extension/blob/main/src/core/ts/background-scripts/background.js/lib/BergamotTranslatorAPI.ts#L227). The issue here is that you don't want to mix `performance.*`,  `error.*` and `metadata.*`metrics from different translations on the same ping, correct? If so, is possible for you to leverage the existing structure to isolate your pings. 

As I mentioned in my previous comment, you can guarantee that Glean API calls are executed in a block, by calling them inside a synchronous function. 

Now, with this in mind, if you call this synchronous function right after a translation is completed, for example, right after this line https://github.com/mozilla-extensions/bergamot-browser-extension/blob/main/src/core/ts/background-scripts/background.js/lib/BergamotTranslatorAPI.ts#L245, that should fix your problem in a much simpler way than the approach here.

Reply:
Yes, with that guarantee and tests that don't require async functions (as per comments above), we can remove GleanRecorder.

(On a side note, translation is completed when all frames in a tab have been translated, and there are multiple requests to BergamotTranslatorAPI per frame, so the right place to send this telemetry is here: https://github.com/motin/bergamot-browser-extension/blob/678856d2b72731edc1d71df798630e5fbb90f904/src/firefox-infobar-ui/ts/background-scripts/background.js/NativeTranslateUiBroker.ts#L205)